File: 160. Intersection of Two Linked Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
import java.util.*;

public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        // cursors
        ListNode cursorA = headA;
        ListNode cursorB = headB;
        // cursorA: walks down listA then listB
        // cursorB: walks down listB then listA
        // cursorA == cursorB when cursorA in listB and cursorB in listA
        // or they both = null at the end of journey when there's no intersection
        while (cursorA != cursorB) {
            if (cursorA == null) {
                cursorA = headB;
            } else {
                cursorA = cursorA.next;
            }
            if (cursorB == null) {
                cursorB = headA;
            } else {
                cursorB = cursorB.next;
            }
        }
        return cursorA; // either the intersection node or null
    }
}

File: 160. Intersection of Two Linked Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        s = set()
        
        # put each node of A into a set
        A = headA
        while A:
            s.add(A)
            A = A.next
            
        # if a node in B is in set, return it as the intersection
        B = headB
        while B:
            if B in s:
                return B
            B = B.next
    
        # no intersection
        return None